File: src/types/request.ts
export interface KeyValueItem {
  key: string
  value: string
  description?: string
  enabled: boolean
}

export type ThemeMode = 'light' | 'dark' | 'system'
export type ResolvedTheme = 'light' | 'dark'
export type AppLocale = 'en' | 'zh-CN'
export type RequestBodyType = 'json' | 'formdata' | 'raw' | 'binary'
export type AuthType = 'none' | 'bearer' | 'basic' | 'apiKey'
export type RequestTestSource = 'status' | 'header' | 'body'
export type RequestTestOperator = 'equals' | 'contains' | 'exists'
export type ResponseLifecycleState = 'idle' | 'pending' | 'success' | 'http-error' | 'transport-error'
export type RequestExecutionSource = 'live' | 'mock'
export type RequestRedirectPolicy = 'follow' | 'manual' | 'error'
export type RequestProxyMode = 'inherit' | 'off' | 'custom'
export type RequestTabOriginKind = 'resource' | 'replay' | 'scratch' | 'detached'
export type RequestTabPersistenceState = 'saved' | 'unsaved' | 'unbound'
export type RequestTabExecutionState = ResponseLifecycleState
export type RequestKind = 'http' | 'mcp'
export type McpTransportKind = 'http' | 'stdio'
export type McpOperationType = 'initialize' | 'tools.list' | 'tools.call' | 'resources.list' | 'resources.read'

export interface WorkbenchActivitySignal {
  active: boolean
  open: boolean
  dirty: boolean
  running: boolean
  recovered: boolean
  result: RequestTabExecutionState
}

export interface WorkbenchActivitySummary {
  open: number
  dirty: number
  running: number
  recovered: number
}

export interface WorkbenchActivityProjection {
  requests: Record<string, WorkbenchActivitySignal>
  history: Record<string, WorkbenchActivitySignal>
  tabs: Record<string, WorkbenchActivitySignal>
  summary: WorkbenchActivitySummary
}

export interface FormDataFieldSnapshot {
  key: string
  value: string
  enabled: boolean
  kind?: 'text' | 'file'
  fileName?: string
  mimeType?: string
}

export interface RequestMockState {
  enabled: boolean
  status: number
  statusText: string
  contentType: string
  body: string
  headers: KeyValueItem[]
}

export type RequestProxySettings =
  | { mode: 'inherit' }
  | { mode: 'off' }
  | { mode: 'custom'; url: string }

export interface RequestExecutionOptions {
  timeoutMs?: number
  redirectPolicy: RequestRedirectPolicy
  proxy: RequestProxySettings
  verifySsl: boolean
}

export type RequestBodySnapshot =
  | { kind: 'json'; value: string }
  | { kind: 'raw'; value: string; contentType?: string }
  | { kind: 'formData'; fields: FormDataFieldSnapshot[] }
  | { kind: 'binary'; bytesBase64: string; fileName?: string; mimeType?: string }

export interface AuthConfig {
  type: AuthType
  bearerToken: string
  username: string
  password: string
  apiKeyKey: string
  apiKeyValue: string
  apiKeyPlacement: 'header' | 'query'
}

export interface RequestTestDefinition {
  id: string
  name: string
  source: RequestTestSource
  operator: RequestTestOperator
  target?: string
  expected?: string
}

export interface RequestTestResult {
  id: string
  name: string
  passed: boolean
  message: string
}

export interface AssertionResultSet {
  passed: boolean
  results: RequestTestResult[]
}

export interface CompiledRequest {
  protocolKey: string
  method: string
  url: string
  params: KeyValueItem[]
  headers: KeyValueItem[]
  body: RequestBodySnapshot
  auth: AuthConfig
  tests: RequestTestDefinition[]
}

export interface NormalizedResponse {
  status: number
  statusText: string
  elapsedMs: number
  sizeBytes: number
  contentType: string
  body: string
  headers: ResponseHeaderItem[]
  truncated: boolean
}

export interface ExecutionArtifact {
  executionSource: RequestExecutionSource
  executedAtEpochMs?: number
  compiledRequest: CompiledRequest
  normalizedResponse: NormalizedResponse
  assertionResults: AssertionResultSet
}

export interface McpConnectionConfig {
  transport: McpTransportKind
  baseUrl: string
  headers: KeyValueItem[]
  auth: AuthConfig
  sessionId?: string
}

export interface McpInitializeInput {
  clientName: string
  clientVersion: string
  protocolVersion?: string
  capabilities?: Record<string, unknown>
}

export interface McpToolSchemaSnapshot {
  name: string
  title?: string
  description?: string
  inputSchema?: Record<string, unknown>
}

export interface McpResourceContentSnapshot {
  uri: string
  mimeType?: string
  text?: string
  blob?: string
}

export interface McpResourceSnapshot {
  uri: string
  name?: string
  title?: string
  description?: string
  mimeType?: string
}

export interface McpToolsListInput {
  cursor?: string
}

export interface McpResourcesListInput {
  cursor?: string
}

export interface McpResourceReadInput {
  uri: string
  resource?: McpResourceSnapshot
}

export interface McpToolCallInput {
  toolName: string
  arguments: Record<string, unknown>
  schema?: McpToolSchemaSnapshot
}

export type McpOperationInput =
  | { type: 'initialize'; input: McpInitializeInput }
  | { type: 'tools.list'; input: McpToolsListInput }
  | { type: 'tools.call'; input: McpToolCallInput }
  | { type: 'resources.list'; input: McpResourcesListInput }
  | { type: 'resources.read'; input: McpResourceReadInput }

export interface McpRequestDefinition {
  connection: McpConnectionConfig
  operation: McpOperationInput
}

export interface McpExecutionArtifact {
  transport: McpTransportKind
  operation: McpOperationType
  protocolRequest?: Record<string, unknown>
  protocolResponse?: Record<string, unknown>
  selectedTool?: McpToolSchemaSnapshot
  cachedTools?: McpToolSchemaSnapshot[]
  selectedResource?: McpResourceSnapshot
  cachedResources?: McpResourceSnapshot[]
  resourceContents?: McpResourceContentSnapshot[]
  sessionId?: string
  errorCategory?: 'transport' | 'session' | 'tool-call' | 'initialize' | 'tool_execution'
}

export type HistoryRequestSnapshot = Omit<SendRequestPayload, 'body' | 'bodyType'> & {
  body: string | RequestBodySnapshot
  bodyType?: RequestBodyType
}

export interface RequestPreset {
  id: string
  requestKind?: RequestKind
  mcp?: McpRequestDefinition
  name: string
  description?: string
  tags?: string[]
  method: string
  url: string
  workspaceId?: string
  collectionId?: string
  collectionName?: string
  params?: KeyValueItem[]
  headers?: KeyValueItem[]
  body?: string
  bodyType?: RequestBodyType
  bodyDefinition?: RequestBodySnapshot
  bodyContentType?: string
  formDataFields?: FormDataFieldSnapshot[]
  binaryFileName?: string
  binaryMimeType?: string
  auth?: Partial<AuthConfig>
  tests?: RequestTestDefinition[]
  mock?: RequestMockState
  executionOptions?: RequestExecutionOptions
}

export interface RequestCollection {
  id: string
  name: string
  expanded: boolean
  requests: RequestPreset[]
}

export interface HistoryItem {
  id: string
  name: string
  method: string
  time: string
  status: number
  url: string
  requestId?: string
  executedAtEpochMs?: number
  statusText?: string
  elapsedMs?: number
  sizeBytes?: number
  contentType?: string
  truncated?: boolean
  responseHeaders?: ResponseHeaderItem[]
  responsePreview?: string
  requestSnapshot?: HistoryRequestSnapshot
  executionSource?: RequestExecutionSource
  mcpArtifact?: McpExecutionArtifact
  mcpSummary?: {
    operation: McpOperationType
    transport: McpTransportKind
    errorCategory?: McpExecutionArtifact['errorCategory']
    toolName?: string
    resourceUri?: string
    sessionId?: string
  }
}

export interface ResponseHeaderItem {
  key: string
  value: string
}

export interface ResponseState {
  requestKind?: RequestKind
  mcpArtifact?: McpExecutionArtifact
  responseBody: string
  status: number
  statusText: string
  time: string
  size: string
  headers: ResponseHeaderItem[]
  contentType: string
  requestMethod: string
  requestUrl: string
  testResults: RequestTestResult[]
  state?: ResponseLifecycleState
  stale?: boolean
  executionSource?: RequestExecutionSource
}

export interface EnvironmentPreset {
  id: string
  name: string
  variables: KeyValueItem[]
}

export interface RequestTabOrigin {
  kind: RequestTabOriginKind
  requestId?: string
  historyItemId?: string
}

export interface RequestTabState {
  id: string
  requestKind?: RequestKind
  mcp?: McpRequestDefinition
  requestId?: string
  origin?: RequestTabOrigin
  persistenceState?: RequestTabPersistenceState
  executionState?: RequestTabExecutionState
  name: string
  description: string
  tags: string[]
  collectionName: string
  collectionId?: string
  method: string
  url: string
  params: KeyValueItem[]
  headers: KeyValueItem[]
  body: string
  bodyType: RequestBodyType
  bodyDefinition?: RequestBodySnapshot
  bodyContentType?: string
  formDataFields?: FormDataFieldSnapshot[]
  binaryFileName?: string
  binaryMimeType?: string
  auth: AuthConfig
  tests: RequestTestDefinition[]
  mock?: RequestMockState
  executionOptions?: RequestExecutionOptions
  response: ResponseState
  isSending: boolean
  isDirty?: boolean
}

export interface SendRequestPayload {
  tabId: string
  requestKind?: RequestKind
  mcp?: McpRequestDefinition
  requestId?: string
  name: string
  description: string
  tags: string[]
  collectionName: string
  method: string
  url: string
  params: KeyValueItem[]
  headers: KeyValueItem[]
  body: string
  bodyType: RequestBodyType
  bodyDefinition?: RequestBodySnapshot
  bodyContentType?: string
  formDataFields?: FormDataFieldSnapshot[]
  binaryFileName?: string
  binaryMimeType?: string
  auth: AuthConfig
  tests: RequestTestDefinition[]
  mock?: RequestMockState
  executionOptions?: RequestExecutionOptions
}

export interface WorkspaceSnapshot {
  locale: AppLocale
  themeMode: ThemeMode
  activeEnvironmentId: string
  environments: EnvironmentPreset[]
  collections: RequestCollection[]
  historyItems: HistoryItem[]
  openTabs: RequestTabState[]
  activeTabId: string
}

export interface WorkspaceSummary {
  id: string
  name: string
  description?: string
  sourceTemplateId?: string
}

export interface WorkspaceSessionSnapshot {
  activeEnvironmentId?: string
  openTabs: RequestTabState[]
  activeTabId?: string
}

File: src-tauri/src/models/request.rs
use serde::{Deserialize, Serialize};

use crate::models::app::HistoryItemDto;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum RequestRedirectPolicyDto {
    Follow,
    Manual,
    Error,
}

impl Default for RequestRedirectPolicyDto {
    fn default() -> Self {
        Self::Follow
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum RequestProxyModeDto {
    Inherit,
    Off,
    Custom,
}

impl Default for RequestProxyModeDto {
    fn default() -> Self {
        Self::Inherit
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct RequestProxySettingsDto {
    #[serde(default)]
    pub mode: RequestProxyModeDto,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub url: Option<String>,
}

impl Default for RequestProxySettingsDto {
    fn default() -> Self {
        Self {
            mode: RequestProxyModeDto::Inherit,
            url: None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct RequestExecutionOptionsDto {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub timeout_ms: Option<u64>,
    #[serde(default)]
    pub redirect_policy: RequestRedirectPolicyDto,
    #[serde(default)]
    pub proxy: RequestProxySettingsDto,
    #[serde(default = "default_verify_ssl")]
    pub verify_ssl: bool,
}

impl Default for RequestExecutionOptionsDto {
    fn default() -> Self {
        Self {
            timeout_ms: None,
            redirect_policy: RequestRedirectPolicyDto::Follow,
            proxy: RequestProxySettingsDto::default(),
            verify_ssl: true,
        }
    }
}

fn default_verify_ssl() -> bool {
    true
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct KeyValueItemDto {
    pub key: String,
    pub value: String,
    #[serde(default)]
    pub description: String,
    #[serde(default = "default_enabled")]
    pub enabled: bool,
}

fn default_enabled() -> bool {
    true
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AuthConfigDto {
    pub r#type: String,
    pub bearer_token: String,
    pub username: String,
    pub password: String,
    pub api_key_key: String,
    pub api_key_value: String,
    pub api_key_placement: String,
}

impl Default for AuthConfigDto {
    fn default() -> Self {
        Self {
            r#type: "none".to_string(),
            bearer_token: String::new(),
            username: String::new(),
            password: String::new(),
            api_key_key: "X-API-Key".to_string(),
            api_key_value: String::new(),
            api_key_placement: "header".to_string(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct RequestTestDefinitionDto {
    pub id: String,
    pub name: String,
    pub source: String,
    pub operator: String,
    #[serde(default)]
    pub target: String,
    #[serde(default)]
    pub expected: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct RequestAssertionResultDto {
    pub id: String,
    pub name: String,
    pub passed: bool,
    pub message: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct FormDataFieldDto {
    pub key: String,
    #[serde(default)]
    pub value: String,
    #[serde(default)]
    pub enabled: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub kind: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub file_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mime_type: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct RequestMockStateDto {
    #[serde(default)]
    pub enabled: bool,
    #[serde(default)]
    pub status: u16,
    #[serde(default)]
    pub status_text: String,
    #[serde(default)]
    pub content_type: String,
    #[serde(default)]
    pub body: String,
    #[serde(default)]
    pub headers: Vec<KeyValueItemDto>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "camelCase")]
pub enum RequestBodyDto {
    Json {
        value: String,
    },
    Raw {
        value: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        content_type: Option<String>,
    },
    FormData {
        fields: Vec<FormDataFieldDto>,
    },
    Binary {
        bytes_base64: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        file_name: Option<String>,
        #[serde(skip_serializing_if = "Option::is_none")]
        mime_type: Option<String>,
    },
}

impl Default for RequestBodyDto {
    fn default() -> Self {
        Self::Json {
            value: String::new(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct SendRequestPayloadDto {
    pub workspace_id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub request_kind: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mcp: Option<McpRequestDefinitionDto>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub active_environment_id: Option<String>,
    pub tab_id: String,
    pub request_id: Option<String>,
    pub name: String,
    pub description: String,
    pub tags: Vec<String>,
    pub collection_name: String,
    pub method: String,
    pub url: String,
    pub params: Vec<KeyValueItemDto>,
    pub headers: Vec<KeyValueItemDto>,
    pub body: RequestBodyDto,
    pub auth: AuthConfigDto,
    #[serde(default)]
    pub tests: Vec<RequestTestDefinitionDto>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mock: Option<RequestMockStateDto>,
    #[serde(default)]
    pub execution_options: RequestExecutionOptionsDto,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct McpStdioConnectionConfigDto {
    #[serde(default)]
    pub command: String,
    #[serde(default)]
    pub args: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cwd: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct McpConnectionConfigDto {
    #[serde(default)]
    pub transport: String,
    #[serde(default)]
    pub base_url: String,
    #[serde(default)]
    pub headers: Vec<KeyValueItemDto>,
    #[serde(default)]
    pub auth: AuthConfigDto,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub session_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stdio: Option<McpStdioConnectionConfigDto>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct McpInitializeInputDto {
    #[serde(default)]
    pub client_name: String,
    #[serde(default)]
    pub client_version: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub protocol_version: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub capabilities: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct McpToolSchemaSnapshotDto {
    #[serde(default)]
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub input_schema: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct McpResourceSnapshotDto {
    #[serde(default)]
    pub uri: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mime_type: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct McpResourceContentSnapshotDto {
    #[serde(default)]
    pub uri: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mime_type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub text: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub blob: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct McpToolsListInputDto {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cursor: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct McpResourcesListInputDto {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cursor: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct McpResourceReadInputDto {
    #[serde(default)]
    pub uri: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub resource: Option<McpResourceSnapshotDto>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct McpToolCallInputDto {
    #[serde(default)]
    pub tool_name: String,
    #[serde(default)]
    pub arguments: serde_json::Value,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub schema: Option<McpToolSchemaSnapshotDto>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct McpPromptArgumentSnapshotDto {
    #[serde(default)]
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub required: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct McpPromptSnapshotDto {
    #[serde(default)]
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub arguments: Option<Vec<McpPromptArgumentSnapshotDto>>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct McpRootSnapshotDto {
    #[serde(default)]
    pub uri: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct McpPromptsListInputDto {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cursor: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct McpPromptGetInputDto {
    #[serde(default)]
    pub prompt_name: String,
    #[serde(default)]
    pub arguments: serde_json::Value,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub prompt: Option<McpPromptSnapshotDto>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum McpOperationInputDto {
    #[serde(rename = "initialize")]
    Initialize { input: McpInitializeInputDto },
    #[serde(rename = "tools.list", alias = "toolsList")]
    ToolsList { input: McpToolsListInputDto },
    #[serde(rename = "tools.call", alias = "toolsCall")]
    ToolsCall { input: McpToolCallInputDto },
    #[serde(rename = "resources.list", alias = "resourcesList")]
    ResourcesList { input: McpResourcesListInputDto },
    #[serde(rename = "resources.read", alias = "resourcesRead")]
    ResourcesRead { input: McpResourceReadInputDto },
    #[serde(rename = "prompts.list", alias = "promptsList")]
    PromptsList { input: McpPromptsListInputDto },
    #[serde(rename = "prompts.get", alias = "promptsGet")]
    PromptsGet { input: McpPromptGetInputDto },
}

impl Default for McpOperationInputDto {
    fn default() -> Self {
        Self::Initialize {
            input: McpInitializeInputDto::default(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct McpRequestDefinitionDto {
    pub connection: McpConnectionConfigDto,
    pub operation: McpOperationInputDto,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub roots: Option<Vec<McpRootSnapshotDto>>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct SendMcpRequestPayloadDto {
    pub workspace_id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub active_environment_id: Option<String>,
    pub tab_id: String,
    pub request_id: Option<String>,
    pub name: String,
    pub description: String,
    pub tags: Vec<String>,
    pub collection_name: String,
    pub request_kind: String,
    pub mcp: McpRequestDefinitionDto,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct AssertionResultSetDto {
    #[serde(default)]
    pub passed: bool,
    #[serde(default)]
    pub results: Vec<RequestAssertionResultDto>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct ResponseHeaderItemDto {
    pub key: String,
    pub value: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct CompiledRequestDto {
    #[serde(default)]
    pub protocol_key: String,
    #[serde(default)]
    pub method: String,
    #[serde(default)]
    pub url: String,
    #[serde(default)]
    pub params: Vec<KeyValueItemDto>,
    #[serde(default)]
    pub headers: Vec<KeyValueItemDto>,
    #[serde(default)]
    pub body: RequestBodyDto,
    #[serde(default)]
    pub auth: AuthConfigDto,
    #[serde(default)]
    pub tests: Vec<RequestTestDefinitionDto>,
    #[serde(default)]
    pub execution_options: RequestExecutionOptionsDto,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct NormalizedResponseDto {
    #[serde(default)]
    pub status: u16,
    #[serde(default)]
    pub status_text: String,
    #[serde(default)]
    pub elapsed_ms: u64,
    #[serde(default)]
    pub size_bytes: usize,
    #[serde(default)]
    pub content_type: String,
    #[serde(default)]
    pub body: String,
    #[serde(default)]
    pub headers: Vec<ResponseHeaderItemDto>,
    #[serde(default)]
    pub truncated: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct ExecutionArtifactDto {
    #[serde(default = "default_execution_source")]
    pub execution_source: String,
    #[serde(default)]
    pub executed_at_epoch_ms: u64,
    #[serde(default)]
    pub compiled_request: CompiledRequestDto,
    #[serde(default)]
    pub normalized_response: NormalizedResponseDto,
    #[serde(default)]
    pub assertion_results: AssertionResultSetDto,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct McpExecutionArtifactDto {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stderr_summary: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub failure_phase: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub session_state: Option<String>,
    #[serde(default)]
    pub transport: String,
    #[serde(default)]
    pub operation: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub protocol_request: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub protocol_response: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub selected_tool: Option<McpToolSchemaSnapshotDto>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cached_tools: Option<Vec<McpToolSchemaSnapshotDto>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub selected_resource: Option<McpResourceSnapshotDto>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cached_resources: Option<Vec<McpResourceSnapshotDto>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub resource_contents: Option<Vec<McpResourceContentSnapshotDto>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub selected_prompt: Option<McpPromptSnapshotDto>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cached_prompts: Option<Vec<McpPromptSnapshotDto>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub roots: Option<Vec<McpRootSnapshotDto>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub session_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error_category: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SendRequestResultDto {
    pub request_method: String,
    pub request_url: String,
    pub status: u16,
    pub status_text: String,
    pub elapsed_ms: u64,
    pub size_bytes: usize,
    pub content_type: String,
    pub response_body: String,
    pub headers: Vec<ResponseHeaderItemDto>,
    pub truncated: bool,
    #[serde(default = "default_execution_source")]
    pub execution_source: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub assertion_results: Option<AssertionResultSetDto>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub execution_artifact: Option<ExecutionArtifactDto>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub history_item: Option<HistoryItemDto>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SendMcpRequestResultDto {
    pub request_method: String,
    pub request_url: String,
    pub status: u16,
    pub status_text: String,
    pub elapsed_ms: u64,
    pub size_bytes: usize,
    pub content_type: String,
    pub response_body: String,
    pub headers: Vec<ResponseHeaderItemDto>,
    pub truncated: bool,
    #[serde(default = "default_execution_source")]
    pub execution_source: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mcp_artifact: Option<McpExecutionArtifactDto>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub history_item: Option<HistoryItemDto>,
}

impl Default for SendMcpRequestResultDto {
    fn default() -> Self {
        Self {
            request_method: "POST".to_string(),
            request_url: String::new(),
            status: 0,
            status_text: "NOT_IMPLEMENTED".to_string(),
            elapsed_ms: 0,
            size_bytes: 0,
            content_type: "application/json".to_string(),
            response_body: String::new(),
            headers: Vec::new(),
            truncated: false,
            execution_source: default_execution_source(),
            mcp_artifact: None,
            history_item: None,
        }
    }
}

impl Default for SendRequestResultDto {
    fn default() -> Self {
        Self {
            request_method: "GET".to_string(),
            request_url: String::new(),
            status: 0,
            status_text: "NOT_IMPLEMENTED".to_string(),
            elapsed_ms: 0,
            size_bytes: 0,
            content_type: "text/plain".to_string(),
            response_body: String::new(),
            headers: Vec::new(),
            truncated: false,
            execution_source: default_execution_source(),
            assertion_results: None,
            execution_artifact: None,
            history_item: None,
        }
    }
}

fn default_execution_source() -> String {
    "live".to_string()
}
